### PR TITLE
feat(threads): add mutex with priority inheritance

### PR DESCRIPTION
**Pull request by elenaf9**
# Description

Add a mutex implementation analogous to the existing `Lock`, with additional priority inheritance.
Unfortunately I don't think we can directly reuse `Lock`, because we need access to the inner `ThreadList`.

The PR also moves all synchronization primitives into a new `sync` submodule. Any objections?

## Issues/PRs references

Issue: #321.

Depends on 
- #370 
- #381
- #452 

## Change checklist

- [ ] I have cleaned up my commit history and squashed fixup commits.
- [x] I have followed the [Coding Conventions](../book/src/coding-conventions.md).
- [x] I have performed a self-review of my own code.
- [x] I have made corresponding changes to the documentation.

### CODE REVIEW

**Comment by elenaf9** [comment]
I've rebased now, and moved the `threading-mutex` example into a `threading-mutex` test instead, analogous to the `thread-lock` test.

**Comment by kaspar030** [comment]
> The PR also moves all synchronization primitives into a new sync submodule. Any objections?

No objections. How hard is it to chop that into another PR that we can merge right away?

**Comment by elenaf9** [comment]
> > The PR also moves all synchronization primitives into a new sync submodule. Any objections?
> 
> No objections. How hard is it to chop that into another PR that we can merge right away?

Not hard at all :slightly_smiling_face: #452.

File: src/riot-rs-threads/src/sync/mutex.rs [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    /// the waiting thread's priority.
+    /// The priority is reset once the mutex is released. This means that a **user can not change a thread's
+    /// priority while it holds the lock**, because it will be changed back after release!
+    ///
+    /// # Panics
+    ///
+    /// Panics if called outside of a thread context.
+    pub fn lock(&self) -> MutexGuard<T> {
+        critical_section::with(|cs| {
+            let state = unsafe { &mut *self.state.get() };
```
[the remark this answers is not in the document]
Comment (elenaf9) [reply src/riot-rs-threads/src/sync/mutex.rs:87]: > I'm not yet convinced that the safety conditions for casting the pointer obtained from that UnsafeCell to a &mut are verified here (as that lock() method only takes a shared reference).

I am not sure if I understand you correctly. Do you mean safe in the sense that no other threads have mutual access at the same time? This is ensured because all mutual access only every happens inside global critical sections.

File: src/riot-rs-threads/src/sync/mutex.rs [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+                        // thread in the waitlist.
+                        Some(waiter_prio) if waiter_prio > *owner_prio => {
+                            // Current mutex owner inherits the priority.
+                            THREADS.with_mut_cs(cs, |mut threads| {
+                                threads.set_priority(*owner_id, waiter_prio)
+                            });
+                        }
+                        _ => {}
+                    }
+                    // Context switch happens here as soon as we leave the critical section.
```
[the remark this answers is not in the document]
Comment (elenaf9) [reply src/riot-rs-threads/src/sync/mutex.rs:109]: I put the comment down here on purpose because when first reading the code, it took me a while to understand that after the `put_current` call a context switch happens, and that's why in the code that then follows it can be assumed that the lock is now acquired (because the context switch back to the thread will only then happen).

**Comment by elenaf9** [comment]
> On another note, when should we use that new mutex? Should we replace our uses of [`embassy_sync::mutex::Mutex`](https://docs.rs/embassy-sync/latest/embassy_sync/mutex/struct.Mutex.html) with it?

Tldr: `embassy_sync::mutex::Mutex` is for usage inside of async tasks, `riot_rs_threads::sync::Mutex` for threads without an embassy executor.

`embassy_sync::mutex::Mutex` is the correct choice if we have (multiple) async tasks, because it only blocks only the waiting task, and other tasks on the same executor can still run. The `riot_rs_threads::sync::Mutex` can a) only be used with a `executor_thread` (otherwise it'd panic when its not inside a thread context, e.g. in case of the `executor-interrupt`), and b) always block the whole thread and thus also all other tasks that are running on the same executor.

The `riot_rs_threads::sync::Mutex`  is for non-async contexts. This could also be realized with the embassy Mutex + `block_on`, but the new Mutex has a smaller overhead and the priority inheritance feature.

File: src/riot-rs-threads/src/sync/mutex.rs [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    ops::{Deref, DerefMut},
+};
+
+use critical_section::CriticalSection;
+use riot_rs_runqueue::{RunqueueId, ThreadId};
+
+use crate::{thread::ThreadState, threadlist::ThreadList, THREADS};
+
+/// A basic mutex with priority inheritance.
+pub struct Mutex<T> {
```
[the remark this answers is not in the document]
Comment (elenaf9) [reply src/riot-rs-threads/src/sync/mutex.rs:12]: What would be the advantage of opting out from the `Sized` bound? We do need it on `Mutex::new` (Embassy and std both have it there as well), and there is not other way to create the Mutex, so `T` must be `Sized`. Am I missing something?

File: src/riot-rs-threads/src/threadlist.rs [hunk 4]
```diff
@@ -35,13 +37,20 @@ impl ThreadList {
                 next = threads.thread_blocklist[usize::from(n)];
             }
             threads.thread_blocklist[usize::from(pid)] = next;
-            match curr {
-                Some(curr) => threads.thread_blocklist[usize::from(curr)] = Some(pid),
-                _ => self.head = Some(pid),
-            }
+            let inherit_priority = match curr {
```
[the remark this answers is not in the document]
Comment (elenaf9) [reply src/riot-rs-threads/src/threadlist.rs:40]: Personal preference for using `match` here, but I changed the second branch to `None` to emphasize that it's not a "catch-all" case.

File: src/riot-rs-threads/src/sync/mutex.rs [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+                }
+            }
+        })
+    }
+}
+
+/// Grants access to the [`Mutex`] inner data.
+///
+/// Dropping the [`MutexGuard`] will unlock the [`Mutex`];
+pub struct MutexGuard<'a, T> {
```
[the remark this answers is not in the document]
Comment (elenaf9) [reply src/riot-rs-threads/src/sync/mutex.rs:170]: Good catch!

File: src/riot-rs-threads/src/sync/mutex.rs [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe { &*self.mutex.inner.get() }
+    }
+}
+
+impl<'a, T> DerefMut for MutexGuard<'a, T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        unsafe { &mut *self.mutex.inner.get() }
```
[the remark this answers is not in the document]
Comment (elenaf9) [reply src/riot-rs-threads/src/sync/mutex.rs:191]: > This is just copy-pasted from the std's implementation, but it seems easy to write

What comment do you mean? I didn't find anything in the std' Mutex deref impl.
But I added a safety comment.

File: src/riot-rs-threads/src/sync/mutex.rs [hunk 7]
```diff
@@ -0,0 +1,205 @@
+use core::{
```
Comment (kaspar030) [inline src/riot-rs-threads/src/sync/mutex.rs:1]:
```suggestion
#![deny(missing_docs)]                                                                                                                                       
#![deny(clippy::pedantic)]

use core::{
```

This also needs a rebase (and adapt to the ariel-os rename).

**Comment by kaspar030** [comment]
This LGTM. Let's update to current main, fix linting issues and merge!

File: src/riot-rs-threads/src/sync/mutex.rs (hunk 7 above)
Comment (elenaf9) [reply src/riot-rs-threads/src/sync/mutex.rs:1]: Why are the `#[deny(...)]` statements needed here? None of the other submodules in `sync` have them.

Comment (kaspar030) [reply src/riot-rs-threads/src/sync/mutex.rs]: New code has: https://github.com/ariel-os/ariel-os/blob/3ae24d8770bc4c79923b70ee629dec9985a1275a/src/ariel-os-threads/src/sync/event.rs#L3-L4

This is basically our @ROMemories bot. It will e.g., notice that the `unsafe impl Sync` doesn't have a safety warning.

Comment (elenaf9) [reply src/riot-rs-threads/src/sync/mutex.rs]: Okay. Added them and fixed the lints.

**Comment by kaspar030** [comment]
Thanks a lot @elenaf9!
